source/h1_stream.c

```diff
@@ -217,9 +286,79 @@ struct aws_h1_stream *aws_h1_stream_new_request_handler(const struct aws_http_re
      * with the lock here. */
     stream->base.id = aws_http_connection_get_next_stream_id(options->server_connection);
 
+    /* Request-handler (server) streams don't need user to call activate() on them.
+     * Since these these streams can only be created on the event-loop thread,
+     * it's not possible for callbacks to fire before the stream pointer is returned.
+     * (Clients must call stream.activate() because they might create a stream on any thread) */
+    stream->synced_data.api_state = AWS_H1_STREAM_API_STATE_ACTIVE;
+
     stream->base.server_data = &stream->base.client_or_server_data.server;
     stream->base.server_data->on_request_done = options->on_request_done;
     aws_atomic_fetch_add(&stream->base.refcount, 1);
 
     return stream;
 }
+
+int aws_h1_stream_send_response(struct aws_h1_stream *stream, struct aws_http_message *response) {
+    struct aws_h1_connection *connection = s_get_h1_connection(stream);
+    int error_code = 0;
+
+    /* Validate the response and cache info that encoder will eventually need.
+     * The encoder_message object will be moved into the stream later while holding the lock */
+    struct aws_h1_encoder_message encoder_message;
+    bool body_headers_ignored = stream->base.request_method == AWS_HTTP_METHOD_HEAD;
+    if (aws_h1_encoder_message_init_from_response(
+            &encoder_message, stream->base.alloc, response, body_headers_ignored, &stream->thread_data.chunk_list)) {
+        error_code = aws_last_error();
+        goto error;
+    }
+
+    bool should_schedule_task = false;
+    { /* BEGIN CRITICAL SECTION */
+        s_stream_lock_synced_data(stream);
```

Comment:
so you're locking stream data, and unlocking connection data in this block, is that correct/expected? Can you document that?

Reply:
Oops, it was an accident to call stream_lock() and connection_unlock(). Changed to call stream_unlock()
But they do the same thing because the connection and all its streams share a single lock.
I added comments anywhere we touch both the stream and connection state while holding a lock.

(We do give locks to the individual streams in H2, which expects multiple simultaneous streams per connection, but it didn't seem worth the complexity here)